packages/users/src/updateUserEmailSettings.ts

```diff
@@ -1,23 +1,62 @@
 'use server';
 
+import { log } from '@charmverse/core/log';
 import { prisma } from '@charmverse/core/prisma-client';
-import { authActionClient } from '@packages/nextjs/actions/actionClient';
+import { deleteSubscriptionByEmail as deleteBeehiivSubscription } from '@packages/beehiiv/deleteSubscriptionByEmail';
+import { registerScout as registerBeehiiv } from '@packages/beehiiv/registerScout';
+import { deleteContact as deleteLoopsContact } from '@packages/loops/client';
+import { registerScout as registerLoops } from '@packages/loops/registerScout';
+import { getPlatform } from '@packages/mixpanel/utils';
+import { isValidEmail } from '@packages/utils/strings';
 
-import { updateUserEmailSettingsSchema } from './updateUserEmailSettingsSchema';
+export async function updateUserEmailSettings({
+  userId,
+  email,
+  sendMarketing,
+  sendTransactionEmails
+}: {
+  userId: string;
+  email: string;
+  sendMarketing: boolean;
+  sendTransactionEmails: boolean;
+}) {
+  if (typeof email !== 'string') {
+    throw new Error('Email is required');
+  }
+  if (!isValidEmail(email)) {
+    throw new Error('Email is invalid');
+  }
 
-export const updateUserEmailSettingsAction = authActionClient
-  .metadata({ actionName: 'update-user-email-settings' })
-  .schema(updateUserEmailSettingsSchema)
-  .action(async ({ parsedInput, ctx }) => {
-    const userId = ctx.session.scoutId;
-    await prisma.scout.update({
-      where: { id: userId },
-      data: {
-        email: parsedInput.email,
-        sendTransactionEmails: parsedInput.sendTransactionEmails,
-        sendMarketing: parsedInput.sendMarketing
-      }
-    });
+  email = email.trim(); // just in case
 
-    return { success: true };
+  const original = await prisma.scout.findUniqueOrThrow({
+    where: {
+      id: userId
+    }
+  });
+  const updatedUser = await prisma.scout.update({
+    where: { id: userId },
+    data: {
+      email,
+      sendTransactionEmails,
+      sendMarketing
+    }
   });
+
+  if (original.email !== updatedUser.email || original.sendMarketing !== updatedUser.sendMarketing) {
+    try {
+      if (!updatedUser.email) {
+        // remove from Loops and Beehiiv
+        await deleteLoopsContact({ email: original.email! });
+        await deleteBeehiivSubscription({ email: original.email! });
+      } else {
+        await registerLoops({ ...updatedUser, oldEmail: original.email }, getPlatform());
+        await registerBeehiiv({ ...updatedUser, oldEmail: original.email });
```

Comment:
This looks a bit weird. When will !updatedUser.email ever be truthy, email is always required. And shouldn't we delete the beehive and loops data when they toggle off the send marketing flag?

Reply:
@Devorein you're right, email is required. i'll just remove that i guess. it wsa copied from charmverse where we allowed no email

the register methods handle sendMarketing = false already. insetad of deleting they just unsubscribe the user